File: python/843-H.py
from typing import List


class Solution:
    def findSecretWord(self, word_list: List[str], master: 'Master') -> None:
        def calcMatch(word0: str, word1: str) -> int:
            return sum(char0 == char1 for char0, char1 in zip(word0, word1))

        for test in range(10):
            n = len(word_list)
            match_matrix = [[0 for j in range(n)] for i in range(n)]
            for i in range(n):
                for j in range(i, n):
                    match = calcMatch(word_list[i], word_list[j])
                    match_matrix[i][j] = match_matrix[j][i] = match

            guess_idx = max((sorted(match_matrix[i], reverse=True), i) for i in range(n))[1]
            guess_word = word_list[guess_idx]

            match = master.guess(guess_word)
            if match == 6:
                return

            word_list = [word_list[j] for j, m in enumerate(match_matrix[guess_idx]) if m == match]

File: c++/843-H.cpp
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * class Master {
 *   public:
 *     int guess(string word);
 * };
 */
class Solution {
public:
    int matchChars(const string& str0, const string& str1) {
        const int n = str0.size();
        int match = 0;
        for (int i = 0; i < n; ++i) {
            match += str0[i] == str1[i];
        }
        return match;
    }

    void findSecretWord(vector<string>& wordlist, Master& master) {
        const int n = wordlist.size();

        vector<vector<int>> match_cnts(n, vector<int>(n, 0));
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                match_cnts[i][j] = matchChars(wordlist[i], wordlist[j]);
            }
        }

        vector<int> candidates(n, 0);
        iota(candidates.begin(), candidates.end(), 0);

        while (true) {
            const int m = candidates.size();

            vector<int> max_cnts(m, 0);
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < m; ++j) {
                    if (i == j) continue;
                    max_cnts[i] = max(max_cnts[i], match_cnts[candidates[i]][candidates[j]]);
                }
            }

            int pick = candidates[max_element(max_cnts.begin(), max_cnts.end()) - max_cnts.begin()];
            auto match = master.guess(wordlist[pick]);
            if (match == wordlist[pick].size()) return;

            vector<int> n_candidates;
            for (int i = 0; i < m; ++i) {
                if (match_cnts[candidates[i]][pick] == match) {
                    n_candidates.push_back(candidates[i]);
                }
            }

            candidates = move(n_candidates);
        }
    }
};